### PR TITLE
coprocessor/dag/expr: implement SubDurationAndDuration

### PR DESCRIPTION
**Pull request by caohe**
Signed-off-by: caohe <caohe9603@gmail.com>

## What have you changed? (mandatory)

- Implement builtin function: SubDurationAndDuration in coprocessor.

## What are the type of the changes? (mandatory)

- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)

- Unit test

## Does this PR affect documentation (docs) update? (mandatory)

- No

## Does this PR affect tidb-ansible update? (mandatory)

- No

## Refer to a related PR or issue link (optional)

- #3275

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -418,6 +418,25 @@ impl ScalarFunc {
...
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, MyDuration>>> {
+        let d0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
+        let d1: Cow<'a, MyDuration> = try_opt!(self.children[1].eval_duration(ctx, row));
+        let diff = match d0.to_nanos().checked_sub(d1.to_nanos()) {
+            Some(result) => result,
+            None => return Err(box_err!("sub duration {} from duration {} error", d1, d0)),
+        };
+        let res = match MyDuration::from_nanos(diff, d0.fsp().max(d1.fsp()) as i8) {
```
Comment (siddontang) [inline src/coprocessor/dag/expr/builtin_time.rs]: can we use `?` instead of `match`?

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -418,6 +418,25 @@ impl ScalarFunc {
...
+
+    #[inline]
+    pub fn sub_duration_and_duration<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, MyDuration>>> {
+        let d0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
+        let d1: Cow<'a, MyDuration> = try_opt!(self.children[1].eval_duration(ctx, row));
+        let diff = match d0.to_nanos().checked_sub(d1.to_nanos()) {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Could implement an additional function for `MyDuration`?

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 3]
```diff
@@ -418,6 +418,25 @@ impl ScalarFunc {
...
+            Some(result) => result,
+            None => return Err(box_err!("sub duration {} from duration {} error", d1, d0)),
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Seems we should return an error with code `ERR_DATA_OUT_OF_RANGE`

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 1 above)
Comment (caohe) [reply src/coprocessor/dag/expr/builtin_time.rs]: @siddontang PTAL.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 3 above)
Comment (caohe) [reply src/coprocessor/dag/expr/builtin_time.rs]: @AndreMouche PTAL.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (caohe) [reply src/coprocessor/dag/expr/builtin_time.rs]: @AndreMouche I find that `checked_sub()` for `MyDuration` is being implemented in #4147. Can I also implement one in my way?

Comment (AndreMouche) [reply src/coprocessor/dag/expr/builtin_time.rs]: Both are ok for me: you could review #4147 or  implemented one in your way

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 4]
```diff
@@ -418,6 +418,27 @@ impl ScalarFunc {
...
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, MyDuration>>> {
+        let d0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
+        let d1: Cow<'a, MyDuration> = try_opt!(self.children[1].eval_duration(ctx, row));
+        let diff = match d0.to_nanos().checked_sub(d1.to_nanos()) {
+            Some(result) => result,
+            None => {
+                return Err(Error::Eval(
+                    format!("sub duration {} from duration {} error", d1, d0),
+                    ERR_DATA_OUT_OF_RANGE,
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Could we use the following instead?
```
 Error::overflow("DURATION", &format!("({} - {})", &d0, &d1));
```

Comment (caohe) [reply src/coprocessor/dag/expr/builtin_time.rs]: @AndreMouche  OK. PTAL

**Comment by AndreMouche** [comment]
@zz-jason @rleungx  PTAL

**Pull request merged**